### ISSUE TITLE
pod-reflector can stop working and doesn't recover

### ISSUE THREAD

**Issue by minrk** [jupyterhub/kubespawner#209]
We recently had an outage on mybinder.org, *likely* due to the pod reflector stopping receiving events. We should check for bugs in this logic, and perhaps add some more proactive "restart if suspicious" logic to ensure the pod reflector isn't allowed to be stuck in this state.

**Comment by choldgraf** [comment]
Note that this behavior is popping up again today...we should figure out what's going on with the reflector

**Comment by yuvipanda** [comment]
Today's outage is from the event reflector making about 4k requests per minute to the k8s API server, so I think is different but related.

**Comment by cbanek** [comment]
I believe I've been seeing something similar to this as well.

What happens for me is that the hub seems to be working well, and allows users to log in and everything is fine.  Then something happens (seemingly randomly, not under load, if anything it seems to happen under an absence of load) and no new users can start labs.  My load is very small (maybe half a dozen users) and we're definitely not hitting some kind of k8s API throttling as far as I can tell.

Suddenly the k8s pod for the new lab container spawns, and comes up fine, but no traffic is routed to it.  The user is presented with an endless spinner, then a 500 after it gives up.  Looking at the jupyterhub logs, the hub doesn't seem to think the lab container has ever come up, but it has.  I've turned on debug logging, and I don't see any additional info that seems relevant.

I get a number of log entries from the hub like this:

2018-07-11 21:50:49.000 MST
[I 2018-07-12 04:50:49.043 JupyterHub base:722] Pending spawn for lsst-epo-guest-1 didn't finish in 10.0 seconds

Then it gives tornado exceptions after a while:

E  [E 2018-07-12 04:50:50.412 JupyterHub user:427] Unhandled error starting lsst-epo-guest-1's server: pod/jupyter-lsst-2depo-2dguest-2d1 did not start in 300 seconds!
 
E  [E 2018-07-12 04:50:50.472 JupyterHub gen:914] Exception in Future <tornado.concurrent.Future object at 0x7f62990a8dd8> after timeout
 
This error repeats a number of times until I'm getting:

E  [E 2018-07-12 04:50:54.166 JupyterHub base:711] Preventing implicit spawn for lsst-epo-guest-1 because last spawn failed: pod/jupyter-lsst-2depo-2dguest-2d1 did not start in 300 seconds!
 
E  [E 2018-07-12 04:50:54.166 JupyterHub web:1591] Uncaught exception GET /hub/user/lsst-epo-guest-1/ (10.128.0.3)
 
Then I get the 500:
2018-07-11 21:50:54.000 MST
[E 2018-07-12 04:50:54.172 JupyterHub log:122] 500 GET /hub/user/lsst-epo-guest-1/ (lsst-epo-guest-1@10.128.0.3) 8.17ms

Looking into the sqlite database at the spawners table, I see that I have normal rows with a server_id field set for accounts that have logged in, but no server_id field for the ones that can't:

7|7||{"pod_name": "jupyter-lsst-2depo-2dguest-2d1"}||2018-07-12 04:50:50.412767

There is also nothing in the state field of the users table, where the other users that work have an empty JSON document "{}":

7|lsst-epo-guest-1|0|2018-07-12 04:50:50.412767|fe37f12ebcf64113b0fcb4baadc1b117||

The system never recovers from this (I've waited hours/days before), and no further users can log in.  Users that have started their labs before this has happened work normally, even after new users stop being able to start labs.

My setup:
Running jupyterhub on kubernetes on GKE using the helm charts with the following command on a fresh environment:

helm install jupyterhub/jupyterhub \
    --version=v0.7-e38c0a9 \
    --name=jupyterhub \
    --values=hub-config.yaml \
    --wait

This has also happened with version v0.6 of the helm charts as well, I was hoping a newer version of things might help, but so far no dice.
